File: src/services/subscribe.ts
/**
 * Async wrapper for the HEBBS SubscribeService gRPC methods.
 */

import type { Metadata } from '@grpc/grpc-js';
import { mapGrpcError } from '../errors.js';
import { grpcUnary, protoToMemory } from '../proto.js';
import type { SubscribePush } from '../types.js';

/* eslint-disable @typescript-eslint/no-explicit-any */

/**
 * Handle for an active HEBBS subscription stream.
 *
 * Use as an async iterator to receive pushes, and call feed() to send text.
 *
 * ```ts
 * const sub = await client.subscribe({ entityId: 'user-123' });
 * await sub.feed('Tell me about user preferences');
 * for await (const push of sub) {
 *   console.log(push.memory.content);
 * }
 * await sub.close();
 * ```
 */
export class Subscription implements AsyncIterable<SubscribePush> {
  private closed = false;

  constructor(
    private readonly _subscriptionId: number,
    private readonly stream: AsyncGenerator<SubscribePush>,
    private readonly feedStub: any,
    private readonly metadata: Metadata,
    private readonly grpcCall: any,
    private readonly tenantId?: string,
  ) {}

  get subscriptionId(): number {
    return this._subscriptionId;
  }

  async feed(text: string): Promise<void> {
    const req: any = {
      subscriptionId: this._subscriptionId,
      text,
    };
    if (this.tenantId) req.tenantId = this.tenantId;

    try {
      await grpcUnary<any>((cb) => this.feedStub.feed(req, this.metadata, cb));
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;

    const req: any = { subscriptionId: this._subscriptionId };
    if (this.tenantId) req.tenantId = this.tenantId;

    try {
      await grpcUnary<any>((cb) =>
        this.feedStub.closeSubscription(req, this.metadata, cb),
      );
    } catch {
      // Best-effort close, swallow errors
    }

    if (this.grpcCall && typeof this.grpcCall.cancel === 'function') {
      this.grpcCall.cancel();
    }
  }

  /**
   * Collect pushes for up to `timeoutMs` milliseconds.
   *
   * Returns as soon as the stream ends, the timeout expires, or
   * `maxPushes` have been collected (whichever comes first).
   */
  async listen(timeoutMs: number = 5000, maxPushes?: number): Promise<SubscribePush[]> {
    const pushes: SubscribePush[] = [];
    const deadline = Date.now() + timeoutMs;
    const iter = this.stream;

    while (maxPushes === undefined || pushes.length < maxPushes) {
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;

      const result = await Promise.race([
        iter.next(),
        new Promise<{ done: true; value: undefined }>((resolve) =>
          setTimeout(() => resolve({ done: true, value: undefined }), remaining),
        ),
      ]);

      if (result.done) break;
      pushes.push(result.value);
    }

    return pushes;
  }

  [Symbol.asyncIterator](): AsyncIterator<SubscribePush> {
    return this.stream;
  }
}

export class SubscribeService {
  constructor(
    private readonly stub: any,
    private readonly metadata: Metadata,
    private readonly tenantId?: string,
  ) {}

  async subscribe(
    entityId?: string,
    confidenceThreshold: number = 0.5,
  ): Promise<Subscription> {
    const req: any = { confidenceThreshold };
    if (entityId) req.entityId = entityId;
    if (this.tenantId) req.tenantId = this.tenantId;

    let grpcCall: any;
    let subId: number;

    try {
      grpcCall = this.stub.subscribe(req, this.metadata);
      const handshake = await new Promise<any>((resolve, reject) => {
        let resolved = false;
        grpcCall.on('data', (msg: any) => {
          if (!resolved) {
            resolved = true;
            resolve(msg);
          }
        });
        grpcCall.on('error', (err: any) => {
          if (!resolved) {
            resolved = true;
            reject(err);
          }
        });
        grpcCall.on('end', () => {
          if (!resolved) {
            resolved = true;
            reject(new Error('Stream ended before handshake'));
          }
        });
      });

      subId =
        handshake.subscriptionId ?? handshake.subscription_id ?? 0;
    } catch (e) {
      throw mapGrpcError(e);
    }

    const metadata = this.metadata;
    const dataStream = createDataStream(grpcCall);

    return new Subscription(
      subId,
      dataStream,
      this.stub,
      metadata,
      grpcCall,
      this.tenantId,
    );
  }
}

async function* createDataStream(
  grpcCall: any,
): AsyncGenerator<SubscribePush> {
  const buffer: any[] = [];
  let resolve: (() => void) | null = null;
  let done = false;
  let streamError: Error | null = null;

  grpcCall.on('data', (msg: any) => {
    if (msg.memory) {
      buffer.push(msg);
      if (resolve) {
        const r = resolve;
        resolve = null;
        r();
      }
    }
  });

  grpcCall.on('error', (err: any) => {
    streamError = err;
    done = true;
    if (resolve) {
      const r = resolve;
      resolve = null;
      r();
    }
  });

  grpcCall.on('end', () => {
    done = true;
    if (resolve) {
      const r = resolve;
      resolve = null;
      r();
    }
  });

  while (true) {
    if (buffer.length > 0) {
      const msg = buffer.shift()!;
      yield {
        subscriptionId: msg.subscriptionId ?? msg.subscription_id ?? 0,
        memory: protoToMemory(msg.memory),
        confidence: msg.confidence ?? 0,
        pushTimestampUs: msg.pushTimestampUs ?? msg.push_timestamp_us ?? 0,
        sequenceNumber: msg.sequenceNumber ?? msg.sequence_number ?? 0,
      };
      continue;
    }

    if (done) {
      if (streamError) throw mapGrpcError(streamError);
      return;
    }

    await new Promise<void>((r) => {
      resolve = r;
    });
  }
}

/* eslint-enable @typescript-eslint/no-explicit-any */

File: src/hebbs/services/subscribe.py
"""Async wrapper for the HEBBS SubscribeService gRPC methods."""

from __future__ import annotations

import asyncio
from typing import Any, AsyncIterator

from hebbs._generated import hebbs_pb2, hebbs_pb2_grpc
from hebbs.exceptions import _map_grpc_error
from hebbs.services.memory import _proto_to_memory
from hebbs.types import SubscribePush


class Subscription:
    """Handle for an active HEBBS subscription stream.

    Use as an async iterator to receive pushes, and call feed() to send text.
    """

    def __init__(
        self,
        subscription_id: int,
        stream: AsyncIterator,
        feed_stub: hebbs_pb2_grpc.SubscribeServiceStub,
        tenant_id: str | None = None,
        grpc_call: Any = None,
    ) -> None:
        self._subscription_id = subscription_id
        self._stream = stream
        self._feed_stub = feed_stub
        self._tenant_id = tenant_id
        self._grpc_call = grpc_call

    @property
    def subscription_id(self) -> int:
        return self._subscription_id

    async def feed(self, text: str) -> None:
        req = hebbs_pb2.FeedRequest(
            subscription_id=self._subscription_id,
            text=text,
        )
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        try:
            await self._feed_stub.Feed(req)
        except Exception as e:
            raise _map_grpc_error(e) from e

    async def close(self) -> None:
        req = hebbs_pb2.CloseSubscriptionRequest(
            subscription_id=self._subscription_id,
        )
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        try:
            await self._feed_stub.CloseSubscription(req)
        except Exception:
            pass
        if self._grpc_call is not None and hasattr(self._grpc_call, "cancel"):
            self._grpc_call.cancel()

    async def listen(self, timeout: float = 5.0, max_pushes: int | None = None) -> list[SubscribePush]:
        """Collect pushes for up to ``timeout`` seconds.

        Returns as soon as the stream ends, the timeout expires, or
        ``max_pushes`` have been collected (whichever comes first).
        """
        pushes: list[SubscribePush] = []
        deadline = asyncio.get_event_loop().time() + timeout
        while max_pushes is None or len(pushes) < max_pushes:
            remaining = deadline - asyncio.get_event_loop().time()
            if remaining <= 0:
                break
            try:
                push = await asyncio.wait_for(self.__anext__(), timeout=remaining)
                pushes.append(push)
            except asyncio.TimeoutError:
                break
            except StopAsyncIteration:
                break
        return pushes

    def __aiter__(self) -> AsyncIterator[SubscribePush]:
        return self

    async def __anext__(self) -> SubscribePush:
        try:
            msg = await self._stream.__anext__()
        except StopAsyncIteration:
            raise
        except Exception as e:
            raise _map_grpc_error(e) from e
        return SubscribePush(
            subscription_id=msg.subscription_id,
            memory=_proto_to_memory(msg.memory),
            confidence=msg.confidence,
            push_timestamp_us=msg.push_timestamp_us,
            sequence_number=msg.sequence_number,
        )


class SubscribeServiceClient:
    """Async client for the HEBBS SubscribeService."""

    def __init__(self, stub: hebbs_pb2_grpc.SubscribeServiceStub, tenant_id: str | None = None) -> None:
        self._stub = stub
        self._tenant_id = tenant_id

    async def subscribe(
        self,
        entity_id: str | None = None,
        confidence_threshold: float = 0.5,
    ) -> Subscription:
        req = hebbs_pb2.SubscribeRequest(confidence_threshold=confidence_threshold)
        if entity_id:
            req.entity_id = entity_id
        if self._tenant_id:
            req.tenant_id = self._tenant_id

        try:
            grpc_call = self._stub.Subscribe(req)
            stream_iter = grpc_call.__aiter__()
            handshake = await stream_iter.__anext__()
        except Exception as e:
            raise _map_grpc_error(e) from e

        sub_id = handshake.subscription_id

        async def _data_stream():
            async for msg in stream_iter:
                if msg.HasField("memory"):
                    yield msg

        return Subscription(
            subscription_id=sub_id,
            stream=_data_stream(),
            feed_stub=self._stub,
            tenant_id=self._tenant_id,
            grpc_call=grpc_call,
        )
